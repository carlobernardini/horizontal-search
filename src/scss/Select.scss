@import 'base';

.orizzonte__filter-select {
    @extend %select;

    &:focus {
        border-color: $color-a;
    }

    &--disabled {
        @extend %select-disabled;
    }
}
