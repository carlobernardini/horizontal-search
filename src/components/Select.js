import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/Filter.scss';
import '../scss/Select.scss';

const Select = ({ disabled, label, onUpdate, options, value }) => (
    <div
        className="orizzonte__filter"
    >
        <div
            className="orizzonte__filter-caption"
        >
            { label }
        </div>
        <select
            className={ classNames('orizzonte__filter-select', {
                'orizzonte__filter-select--disabled': disabled
            }) }
            disabled={ disabled }
            onChange={ (e) => {
                const { value: val } = e.target;
                onUpdate(val);
            }}
            value={ value }
        >
            { options.map((option, i) => (
                <option
                    key={ `${ option.value }.${ i }` }
                    value={ option.value }
                >
                    { option.label || option.value }
                </option>
            )) }
        </select>
    </div>
);

Select.propTypes = {
    /** If the select should be disabled */
    disabled: PropTypes.bool,
    /** Label for this filter section */
    label: PropTypes.string.isRequired,
    /** Internal callback for when select value has changed */
    onUpdate: PropTypes.func,
    /** List of selectable options (value is required) */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string,
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired
        })
    ).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string
    ])
};

Select.defaultProps = {
    disabled: false,
    onUpdate: () => {},
    value: null
};

export default Select;
