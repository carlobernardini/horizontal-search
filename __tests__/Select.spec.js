import React from 'react';
import Select from '../src/components/Select';

const onUpdate = jest.fn();

describe('<Select />', () => {
    it('should render a normal select', () => {
        const expectedValue = 'test';
        const wrapper = shallow(
            <Select
                label="Test select"
                options={ [{
                    label: 'Test value 1',
                    value: 1
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3
                }]}
                onUpdate={ onUpdate }
            />
        );

        wrapper.find('.orizzonte__filter-select').simulate('change', {
            target: {
                value: expectedValue
            }
        });
        expect(onUpdate).toHaveBeenCalledWith(expectedValue);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a disabled select that allows empty value', () => {
        const wrapper = shallow(
            <Select
                label="Test select"
                notSetLabel="None"
                options={ [{
                    label: 'Test value 1',
                    value: 1
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3
                }]}
                disabled
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a select with selected value', () => {
        const wrapper = shallow(
            <Select
                label="Test select"
                options={ [{
                    label: 'Test value 1',
                    value: 1
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3
                }]}
                value={ 3 }
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a grouped select', () => {
        const wrapper = shallow(
            <Select
                label="Test select"
                options={ [{
                    label: 'Test value 1',
                    value: 1
                }, {
                    value: 'Grouped values',
                    children: [{
                        label: 'Test value 2',
                        value: 2
                    }, {
                        label: 'Test value 3',
                        value: 3
                    }]
                }, {
                    value: 4,
                    label: 'Test value 4'
                }]}
                value={ 3 }
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
