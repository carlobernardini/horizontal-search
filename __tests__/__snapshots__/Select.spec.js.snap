// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render a disabled select 1`] = `
<div
  className="orizzonte__filter"
>
  <div
    className="orizzonte__filter-caption"
  >
    Test select
  </div>
  <select
    className="orizzonte__filter-select orizzonte__filter-select--disabled"
    disabled={true}
    onChange={[Function]}
    value=""
  >
    <option
      key="1.0"
      value={1}
    >
      Test value 1
    </option>
    <option
      key="2.1"
      value={2}
    >
      Test value 2
    </option>
    <option
      key="3.2"
      value={3}
    >
      Test value 3
    </option>
  </select>
</div>
`;

exports[`<Select /> should render a normal select 1`] = `
<div
  className="orizzonte__filter"
>
  <div
    className="orizzonte__filter-caption"
  >
    Test select
  </div>
  <select
    className="orizzonte__filter-select"
    disabled={false}
    onChange={[Function]}
    value=""
  >
    <option
      key="1.0"
      value={1}
    >
      Test value 1
    </option>
    <option
      key="2.1"
      value={2}
    >
      Test value 2
    </option>
    <option
      key="3.2"
      value={3}
    >
      Test value 3
    </option>
  </select>
</div>
`;

exports[`<Select /> should render a select with selected value 1`] = `
<div
  className="orizzonte__filter"
>
  <div
    className="orizzonte__filter-caption"
  >
    Test select
  </div>
  <select
    className="orizzonte__filter-select"
    disabled={false}
    onChange={[Function]}
    value={3}
  >
    <option
      key="1.0"
      value={1}
    >
      Test value 1
    </option>
    <option
      key="2.1"
      value={2}
    >
      Test value 2
    </option>
    <option
      key="3.2"
      value={3}
    >
      Test value 3
    </option>
  </select>
</div>
`;
